Playwright: fold functionality of LoginFlow into LoginPage.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to refactor LoginFlow and LoginPage to merge functionality into LoginPage.

Key changes:
- remove LoginFlow layer of abstraction.
- implement methods in LoginPage to serve as replacement of LoginFlow.

Background:

The current abstraction layers of `LoginPage` and `LoginFlow` were carried over from the Selenium suite. It has become clear in recent days that this abstraction is unnecessary and only serves to add too many layers of code.

In this refactored codebase, `LoginPage` will perform all actions of logging into the webapp. Implementation of the login process is separated into two parts:

`baseflow`: this will serve as the action of filling out the username/email and password, then hitting submit.
`login`: this will be the method called by test writers. Credentials are passed into this method and any other preparations are made (such as navigating to the login endpoint).

#### Testing instructions

- [x] tests all pass
  - [x] desktop
  - [x] mobile
- [x] stress test
  - [x] mobile
  - [x] desktop


Closes #56400.

worldomonation committed:
Refactor LoginFlow out of the framework.
- move functionality of LoginFlow into LoginPage.
- most default flows will navigate user to the login page.
- remove `loginFromPopup` and handle popups in the spec.

Fix two tests:
- notifications spec
- reader view spec

Add waits for `continue-as-user` div to be stable.

worldomonation commented on test/e2e/specs/specs-playwright/wp-likes__post-spec.js:
```diff
@@ -71,15 +71,21 @@ describe( DataHelper.createSuiteTitle( 'Likes (Post)' ), function () {
+			// Observe for the new popup in the Promise. Once the Page object
+			// is obtained, execute login steps there and wait for the
+			// popup to close.
+			await Promise.all( [
+				page.on( 'popup', async ( popup ) => {
+					const loginPage = new LoginPage( popup );
+					await loginPage.login( { account: likeUser } );
+					await popup.waitForEvent( 'close' );
+				} ),
+				publishedPostPage.likePost(),
```
Previously, there was a method `loginFromPopup` that would act on the Page object of the popup page. Now that `LoginFlow` is being removed, I have decided against implementing a similar method in `LoginPage`.

It felt odd to have `LoginPage` incorporate the behavior of `LoginFlow` where an instance of LoginPage is created by passing into the constructor one instance of a page when being constructed yet contain methods that accept another instance of a Page object and work its actions on the other instance.

Since the instances of logging into Calypso via a popup is exactly 1 in the entire suite, I decided it was better to handle popups on a case-by-case basis.

dpasque replied:
I definitely agree, but the only part that feels off is the initial `visit` call in a case like this. Is it worth having two methods, like `login` and `visitAndLogin`, with the only difference being the initial `visit` call? Having those two different methods might be useful over time?

worldomonation replied:
> two methods, like login and visitAndLogin

I am on the side that it does not make sense to write a method used by one test step only, but the `visit` embedded within the `login` method is also possibly problematic.

My proposal is this: re-add the method named `loginFromPopup` which omits the `visit` call. This will be used for all tests that require logging in from a popup.

How does that sound?

dpasque replied:
Oh yes, I like that! Good idea!!

WunderBart commented on test/e2e/specs/specs-playwright/wp-invite__new-user.ts:
```diff
@@ -12,7 +16,7 @@ import {
 import { Page } from 'playwright';
 
-describe.skip( DataHelper.createSuiteTitle( `Invite: New User` ), function () {
+describe( DataHelper.createSuiteTitle( `Invite: New User` ), function () {
```
Is it intended to un-skip this one within this PR?

worldomonation replied:
Yep, I added the proper Jest test group headers to the file.

worldomonation replied:
I've triggered an instance of the pre-release tests with this test enabled - if it the test has issues, I will remove this change from the PR and address it separately.

worldomonation replied:
I'll remove this change from this PR, since it isn't explicitly tied to the refactoring of LoginFlow.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +54,113 @@ export class LoginPage {
 		if ( alreadyLoggedIn ) {
-			console.log( 'already logged in, selecting "change account' );
+			const elementHandle = await this.page.waitForSelector( selectors.continueAsUser );
+			await elementHandle.waitForElementState( 'stable' );
```
This part addresses #56400 by waiting for the `continueAsUser` div to finish fading in before proceeding.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +54,113 @@ export class LoginPage {
-		await this.switchAccount();
-
-		// Begin the process of logging in.
+	private async baseflow( { username, password }: LoginCredentials ): Promise< void > {
```
Note the baseflow (which interact with the actual Login page) only accepts the username/email and password pair.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +54,113 @@ export class LoginPage {
 	 */
-	async visit(): Promise< void > {
-		await this.page.goto( getCalypsoURL( 'log-in' ), { waitUntil: 'networkidle' } );
+	async login( credentials: LoginCredentials | TestAccount ): Promise< void > {
```
This is the method that most test writers will interact with, and it allows the test writer to specify two types of credentials.

I kept this behavior from existing `LoginFlow` as it would be quite tedious to retrieve the LoginCredentials pair for test users stored in the configuration within the spec file. I think the convenience of allowing users to specify the test user outweighs any complexity.

One note - whereas `LoginFlow` had the default user of `defaultUser` if a test account was not specified, the refactored code will force test writers to explicitly specify the test account.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +54,113 @@ export class LoginPage {
 	 * @param {string} user Username or email address of the user requesting a link.
 	 */
 	async requestMagicLink( user: string ): Promise< void > {
+		await this.visit();
+		await this.switchAccount();
+
```
Meant to be an all-encapsulating method that performs required steps to get this done.

worldomonation commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -116,29 +114,11 @@ describe(
-
 			it( 'Close account', async function () {
-				const closeAccountFlow = new CloseAccountFlow( newPage );
+				// Magic link only works on production webapp.
+				// Revisit the webapp under test.
+				await page.goto( DataHelper.getCalypsoURL() );
+				const closeAccountFlow = new CloseAccountFlow( page );
```
@dpasque I figured out that all that's required is to redirect the page to the value returned by `getCalypsoURL`. On Pre-Release tests this returns the `wpcalypso.wordpress.com` URL, and once navigated there we can proceed with the account deletion.

Removing the extraneous login step reduces the runtime by however many seconds the login process takes.

Here are the logs confirming that `wordpress.com` is redirected to `wpcalypso.wordpress.com` properly:

```
 navigated to "https://wordpress.com/log-in/link/use
 navigated to "https://wordpress.com/
...
navigating to "https://wpcalypso.wordpress.com/", waiting until "load"
 navigated to "https://wpcalypso.wordpress.com/home/e2eflowtesting1632438661097641.wordpress.com

```

worldomonation replied:
Also, I plan on moving this test out of `Signup: Free` suite and into its own standalone test, so this change will churn again soon.

worldomonation commented:
Stress test passed with the following conditions: `Log in`, then `Switch user`.

[mobile](https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/6748324)
[desktop](https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_desktop/6748336)

fullofcaffeine commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -637,7 +637,7 @@ fun playwrightPrBuildType( targetDevice: String, buildUuid: String ): BuildType
 					export DEBUG=pw:api
 
-					xvfb-run yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% --group=calypso-pr
+					for i in {1..100}; do xvfb-run yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% specs/specs-playwright/wp-media__upload; done
```
I assume this will be reverted before merging?

worldomonation replied:
Yes, this is only for stress testing purposes.

worldomonation replied:
Removed in 94ca9bd42a5816208ce94c9248a11e7ec69a6d57.

fullofcaffeine commented on test/e2e/specs/specs-playwright/wp-media__upload-spec.ts:
```diff
@@ -39,36 +39,36 @@ describe( DataHelper.createSuiteTitle( 'Media: Upload' ), () => {
-				}
-			}
-		} );
+		// it( 'Upload an unsupported file type and see the rejection notice', async function () {
```
Are these examples that are currently failing due to the refactor, and will be eventually worked out before merging? Or what was the reason to comment those out?

worldomonation replied:
This was for stress testing of the changes.

worldomonation replied:
Removed in 94ca9bd42a5816208ce94c9248a11e7ec69a6d57.

worldomonation committed:
Update:
- use correct method name in Signup: Free
- update Signup: WPCC to look for rejection of the login attempt
- modify LoginPage.baseflow to account for rejected credentials.

worldomonation commented:
Rebased and modified some logic around the `LoginPage.baseflow` that required additional stress-testing to validate.

Results are looking good: [desktop](https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_desktop/6816695) and [mobile](https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/6816687).

worldomonation committed:
Stress test

worldomonation committed:
Use `waitForNavigation` with specific URL endpoint

worldomonation committed:
Move the `waitForNavigation` to the `login` method level, not the `baseflow`.
Update newly added tests to use the LoginPage flow.

worldomonation committed:
Remove stress testing code

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +57,126 @@ export class LoginPage {
+			( { username, password } = credentials );
+		}
+
+		await Promise.all( [
+			this.page.waitForNavigation( { waitUntil: 'load' } ),
+			this.baseflow( { username, password } ),
```
(Originally commented on [this commit](https://github.com/Automattic/wp-calypso/pull/56482/commits/429de85b51da08ed80700a4b642add93c8d89f4f), but I've since reverted the approach.)

(I've pasted the comment here for evaluation on the different approaches).

This section has the potential to be problematic if the expected use case of this library shifts.

Currently, every method that calls `baseflow` will redirect to either the Home dashboard or the Reader feature. There currently exist no test that attempt to log in with a `redirect` in the URL parameter that would take the user elsewhere upon successful login.

However, should a test like that be written in the future, the model of behavior codified by these statements will break. At that point, there will need to be a better way to determine whether the user has 1. successfully logged in, and; 2. the page has settled to whatever page they were to be redirected to.

For future reference, `waitForLoadState( 'networkidle' )` is not acceptable. On Simple sites, it adds ~30s to the runtime for no good reason.

Importing a page object (eg. MyHomePage) and waiting on that, or waiting for a specific text string, will also not work as the redirect can be to any Calypso page.

Lastly, a simple `waitForNavigation` is not 100% functional as [this stress test result](https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/6817189) shows.

I am interested in what you think about this though, @dpasque!

dpasque replied:
These failures keep cropping us where that initial redirect (for example from `wordpress.com` to `wordpress.com/home/<site-name>` just keep causing breaks! And that initial redirect definitely seems to be what is causing any of these "eaten-click" kind of failures. 

I know that before I said we should just have people wrap this `login` method in their own url-specific `waitForNavigation`, but I'm not having a change of heart on that. In the current state, almost everyone is landing on the home page, and that initial redirect is still causing a lot of failures.

Here's what I currently think we should do:

1. This current method should take in an optional URL to wait for. 
2. If no URL is provided, we default to the home page URL.
3. We then wait for that specific URL on the navigation.

This should solve the majority case with no changes to the signature, and give us headroom to handle other redirects.

```typescript
interface LoginOptions {
	landingUrl?: string;
}
// ...

	async login(
		credentials: LoginCredentials | TestAccount,
		options?: LoginOptions
	): Promise< void > {
		await this.visit();
		await this.switchAccount();

		let username;
		let password;

		if ( 'account' in credentials ) {
			// Test Account specified. Look for the corresponding username/password
			// combination from the configuration file.
			[ username, password ] = getAccountCredential( credentials.account );
		} else {
			// Regular username/password pair specified - destructure it.
			( { username, password } = credentials );
		}

		const landingUrl = options?.landingUrl
			? options.landingUrl
			: `${ await DataHelper.getCalypsoURL() }/home/**`;

		await Promise.all( [
			this.page.waitForNavigation( { url: landingUrl, waitUntil: 'load' } ),
			this.baseflow( { username, password } ),
		] );
	}
```

worldomonation replied:
I think that makes sense. It allows us to keep the API simple, reducing burden on test writers while also giving us flexibility to handle a unique redirect.

dpasque commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +57,126 @@ export class LoginPage {
+
 		await Promise.all( [
 			this.page.waitForNavigation(),
 			this.page.click( selectors.createAccountLnk ),
```
Nit: small typo! :)

worldomonation replied:
Fixed!

dpasque commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +57,126 @@ export class LoginPage {
 	/**
 	 * Switch account if an user is already logged in.
 	 */
 	private async switchAccount(): Promise< void > {
```
Small optional nit, but can we call this `switchAccountIfNeeded` or something like that? Reading it in the `login` function flow threw me off at first, because I was like "why do we always need to switch account?"

worldomonation replied:
Personally I am not a fan of conditional naming for methods, like `doXYZIfNeeded` or `doXYZIfNecessary`.

I'll try to think of a better name that is not necessarily conditional, but conveys the intent better.

worldomonation replied:
I've renamed it in 977a7843f937196b60541b90fbcf93b91bbe7cf6 to `revealLoginForm`. What do you think? I think it conveys the purpose of the method much better.

dpasque replied:
Good point on the conditional name, and that's a great alternative! I will re-use that idea going forward: you're basically describing the end state that will be ensured - that's brilliant!

dpasque reviewed (commented):
I love these changes so much! I left a couple small optional suggestions, and then a design proposal to handle the redirect flakiness! Let me know what you think, but then this is just about read to go!

worldomonation committed:
Remove LoginFlow page object.

worldomonation committed:
Incorporate suggested changes from @dpasque:
- implement new loginFromPopup method that omits the initial navigation and switch account check.
- extract the credential handling method to a standalone private method.

worldomonation committed:
Rename `switchAccount` to `revealLoginForm`.

worldomonation committed:
Fix tests

dpasque commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -43,75 +61,161 @@ export class LoginPage {
-		await this.page.goto( getCalypsoURL( 'log-in' ), { waitUntil: 'networkidle' } );
+	private async resolveUserCredentials(
+		requestedCredentials: LoginCredentials | TestAccount
+	): Promise< { [ key: string ]: string } > {
```
One small change I missed before: can we change this to return `Promise< LoginCredentials >`?

dpasque replied:
Or optionally, if we want to decouple the typings of what comes in from what resolved credentials look like (like if we wanted to add `site` to the return value or something), we could just make a new interface to represent what is returned!

I just think it would be best to get away from the generic typing 👍

worldomonation replied:
I made the change and TypeScript compiler did not complain and was able to run a spec as expected. Thanks for the heads up!

dpasque reviewed (approved):
One small request, but I love the changes! Nice work! 🎉

worldomonation committed:
Use custom interface as expected return type.
